flax/experimental/nnx/nnx/graph_utils.py

```diff
@@ -860,6 +887,101 @@ def _graph_update_static(
       node_impl.set_key(node, name, value_updates)
 
 
+@tp.overload
+def split(graph_node: A) -> tuple[State, GraphDef[A]]:
+  ...
+
+
+@tp.overload
+def split(
+  graph_node: A, first: filterlib.Filter, /
+) -> tuple[State, GraphDef[A]]:
+  ...
+
+
+@tp.overload
+def split(
+  graph_node: A,
+  first: filterlib.Filter,
+  second: filterlib.Filter,
+  /,
+  *filters: filterlib.Filter,
+) -> tuple[State, tpe.Unpack[tuple[State, ...]], GraphDef[A]]:
+  ...
+
+
+def split(
+  graph_node: A, *filters: filterlib.Filter
+) -> tuple[State, tpe.Unpack[tuple[State, ...]], GraphDef[A]]:
+  state, graphdef, _ = graph_flatten(graph_node)
+
+  if len(filters) == 0:
+    states = (state,)
+  elif len(filters) == 1:
+    states = (state.split(filters[0]),)
+  else:
+    states = state.split(filters[0], filters[1], *filters[2:])
+
+  return *states, graphdef
+
+
+def merge(
+  *states_and_def: tpe.Unpack[
+    tuple[tpe.Unpack[tuple[State, ...]], GraphDef[A]]
+  ],
+) -> A:
+  # TODO: add docstring of example usage
+  *states, graphdef = states_and_def
+  return graphdef.merge(*states)
+
+
+def update(self: A, update: Updates[A], /, *updates: Updates[A]) -> None:
+  updates = (update, *updates)
+
+  def _states_and_moduledef(
+    updates,
+  ) -> tuple[list[State], tp.Optional[A]]:
+    leaves = jax.tree_util.tree_leaves(
+      updates, is_leaf=lambda x: isinstance(x, (GraphDef, State))
+    )
+    states: list[State] = []
+    node: tp.Optional[A] = None
+
+    for leaf in leaves:
+      if is_graph_node(leaf) or isinstance(leaf, GraphDef):
+        if node is not None:
+          raise ValueError(
+            'Expected only one GraphDef or GraphNode in the updates'
+          )
+
+        if is_graph_node(leaf):
+          node = tp.cast(A, leaf)
```

Comment:
As in the other PR, I wonder if you can avoid a `cast` if you make `is_graph_node` a type guard?

Reply:
sadly no its more like a protocol in this case

Reply:
but replace is an `instance` check